File: src-tauri/src/commands/cookie.rs
use crate::{db::AppDb, error::CmdResult, types::Cookie};
use std::collections::HashSet;
use tauri::State;

/// 列出 Cookie（按作用域筛选）
#[tauri::command]
pub async fn list_cookies(
    db: State<'_, AppDb>,
    scope_type: String,
    project_id: Option<i64>,
) -> CmdResult<Vec<Cookie>> {
    let rows = if scope_type == "global" {
        sqlx::query_as::<_, Cookie>(
            "SELECT id, scope_type, project_id, domain, name, value, path, expires_at, http_only, secure, enabled FROM cookies WHERE scope_type='global' ORDER BY id DESC",
        )
        .fetch_all(&db.0)
        .await?
    } else {
        sqlx::query_as::<_, Cookie>(
            "SELECT id, scope_type, project_id, domain, name, value, path, expires_at, http_only, secure, enabled FROM cookies WHERE scope_type='project' AND project_id=? ORDER BY id DESC",
        )
        .bind(project_id)
        .fetch_all(&db.0)
        .await?
    };
    Ok(rows)
}

/// 创建 Cookie
#[tauri::command]
pub async fn create_cookie(
    db: State<'_, AppDb>,
    scope_type: String,
    project_id: Option<i64>,
    domain: String,
    name: String,
    value: String,
    path: String,
) -> CmdResult<Cookie> {
    let row = sqlx::query_as::<_, Cookie>(
        "INSERT INTO cookies (scope_type, project_id, domain, name, value, path) VALUES (?, ?, ?, ?, ?, ?) RETURNING id, scope_type, project_id, domain, name, value, path, expires_at, http_only, secure, enabled",
    )
    .bind(&scope_type)
    .bind(project_id)
    .bind(&domain)
    .bind(&name)
    .bind(&value)
    .bind(&path)
    .fetch_one(&db.0)
    .await?;
    Ok(row)
}

/// 更新 Cookie
#[tauri::command]
pub async fn update_cookie(
    db: State<'_, AppDb>,
    id: i64,
    value: String,
    path: String,
    enabled: i64,
) -> CmdResult<Cookie> {
    let row = sqlx::query_as::<_, Cookie>(
        "UPDATE cookies SET value=?, path=?, enabled=? WHERE id=? RETURNING id, scope_type, project_id, domain, name, value, path, expires_at, http_only, secure, enabled",
    )
    .bind(&value)
    .bind(&path)
    .bind(enabled)
    .bind(id)
    .fetch_one(&db.0)
    .await?;
    Ok(row)
}

/// 删除 Cookie
#[tauri::command]
pub async fn delete_cookie(db: State<'_, AppDb>, id: i64) -> CmdResult<()> {
    sqlx::query("DELETE FROM cookies WHERE id=?")
        .bind(id)
        .execute(&db.0)
        .await?;
    Ok(())
}

/// 获取某域名可用 Cookie（项目级优先覆盖全局同名+同 path）
#[tauri::command]
pub async fn get_cookies_for_domain(
    db: State<'_, AppDb>,
    domain: String,
    project_id: Option<i64>,
) -> CmdResult<Vec<Cookie>> {
    let rows = sqlx::query_as::<_, Cookie>(
        "SELECT id, scope_type, project_id, domain, name, value, path, expires_at, http_only, secure, enabled FROM cookies WHERE domain=? AND enabled=1 AND (scope_type='global' OR (scope_type='project' AND project_id=?)) ORDER BY scope_type DESC, id DESC",
    )
    .bind(&domain)
    .bind(project_id)
    .fetch_all(&db.0)
    .await?;

    let mut seen = HashSet::new();
    let mut deduped = Vec::new();
    for cookie in rows {
        let dedup_key = format!("{}\u{0000}{}", cookie.name, cookie.path);
        if seen.insert(dedup_key) {
            deduped.push(cookie);
        }
    }

    Ok(deduped)
}

File: src/stores/cookie.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { invoke } from '@tauri-apps/api/core'
import type { CookieItem } from '../types'

export const useCookieStore = defineStore('cookie', () => {
  const globalCookies = ref<CookieItem[]>([])
  const projectCookies = ref<CookieItem[]>([])

  async function loadGlobalCookies() {
    // Tauri 2.x #[command] 宏把 Rust snake_case 参数名转为 camelCase IPC key
    globalCookies.value = await invoke<CookieItem[]>('list_cookies', {
      scopeType: 'global',
      projectId: null,
    })
  }

  async function loadProjectCookies(projectId: number) {
    projectCookies.value = await invoke<CookieItem[]>('list_cookies', {
      scopeType: 'project',
      projectId,
    })
  }

  async function createCookie(
    scopeType: string,
    projectId: number | null,
    domain: string,
    name: string,
    value: string,
    path: string,
  ) {
    const c = await invoke<CookieItem>('create_cookie', {
      scopeType,
      projectId,
      domain, name, value, path,
    })
    if (scopeType === 'global') {
      globalCookies.value.push(c)
    } else {
      projectCookies.value.push(c)
    }
    return c
  }

  async function updateCookie(id: number, value: string, path: string, enabled: number) {
    const updated = await invoke<CookieItem>('update_cookie', { id, value, path, enabled })
    const updateList = (list: CookieItem[]) => {
      const idx = list.findIndex(c => c.id === id)
      if (idx !== -1) list[idx] = updated
    }
    updateList(globalCookies.value)
    updateList(projectCookies.value)
    return updated
  }

  async function deleteCookie(id: number) {
    await invoke('delete_cookie', { id })
    globalCookies.value = globalCookies.value.filter(c => c.id !== id)
    projectCookies.value = projectCookies.value.filter(c => c.id !== id)
  }

  return {
    globalCookies,
    projectCookies,
    loadGlobalCookies,
    loadProjectCookies,
    createCookie,
    updateCookie,
    deleteCookie,
  }
})
